src/library_pthread.js

```diff
@@ -445,32 +489,39 @@ var LibraryPThread = {
-    Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.attr }}} + 24) >> 2, threadParams.schedPrio);
-
-    var global_libc = _emscripten_get_global_libc();
-    var global_locale = global_libc + {{{ C_STRUCTS.libc.global_locale }}};
-    Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.locale }}}) >> 2, global_locale);
+    // Worker has been removed from the unused pool, but if the code below throws before it's added to the running pool
+    // it'll be left unaccounted for, so we should protect against that.
+    // Also we don't want to mark a worker as running until it has an associated pthread.
+    try {
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.threadStatus }}} ) >> 2, 0); // threadStatus <- 0, meaning not yet exited.
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.threadExitCode }}} ) >> 2, 0); // threadExitCode <- 0.
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.profilerBlock }}} ) >> 2, 0); // profilerBlock <- 0.
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.detached }}} ) >> 2, threadParams.detached);
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.tsd }}} ) >> 2, tlsMemory); // Init thread-local-storage memory array.
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.tsd_used }}} ) >> 2, 0); // Mark initial status to unused.
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.tid }}} ) >> 2, pthread.threadInfoStruct); // Main thread ID.
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.pid }}} ) >> 2, PROCINFO.pid); // Process ID.
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.attr }}}) >> 2, threadParams.stackSize);
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.stack_size }}}) >> 2, threadParams.stackSize);
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.stack }}}) >> 2, stackHigh);
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.attr }}} + 8) >> 2, stackHigh);
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.attr }}} + 12) >> 2, threadParams.detached);
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.attr }}} + 20) >> 2, threadParams.schedPolicy);
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.attr }}} + 24) >> 2, threadParams.schedPrio);
+      Atomics.store(HEAPU32, (pthread.threadInfoStruct + {{{ C_STRUCTS.pthread.locale }}}) >> 2, global_locale);
 
 #if PTHREADS_PROFILING
-    PThread.createProfilerBlock(pthread.threadInfoStruct);
+      PThread.createProfilerBlock(pthread.threadInfoStruct);
 #endif
-
+    }
+    catch(e) {
```

Comment:
I've checked the standard for Atomics.store and it [can indeed throw](https://www.ecma-international.org/ecma-262/8.0/#sec-validatesharedintegertypedarray). I'd be inclined therefore to keep the try-catch clause. I'm trying to fix the consistency of the worker pool so that we (my company's releases not the least...) can rely on it.

Reply:
> I've checked the standard for Atomics.store and it [can indeed throw](https://www.ecma-international.org/ecma-262/8.0/#sec-validatesharedintegertypedarray).

A "can throw" observation does not yet mean that there should be a try-catch. If we look at the actual conditions that the throws can occur in, those are e.g. that
 - the first parameter to the Atomics op is not a proper SAB (in which case it is a fatal programming error from Emscripten runtime we would never have even been able to run the page at all up to this point),
 - the index parameter is out of bounds (likewise, a fatal programming error, in which case the code should have aborted)

There does not exist any condition where the Atomics.store would throw, that would be up to the developer to handle, but if it does throw, it means that the runtime has a really bad programming bug already, or some other fatal browser or consistency bug has occurred. Those are not scenarios where we should program defensive measures against.